Core/HttpHandlerProxy.cs

```diff
@@ -0,0 +1,151 @@
+        /// </summary>
+        /// <returns>true if the default credentials are used; otherwise false. The default value is false.</returns>
+        public bool UseDefaultCredentials { get; set; }
+
+        public HttpHandlerProxy(IHttpMessageHandlerFactory httpMessageHandlerFactory, string httpClientName)
+        {
+            this.httpClientName = httpClientName;
+            this.httpMessageHandler = httpMessageHandlerFactory?.CreateHandler(httpClientName);
+            this.httpClientHandler = null == httpMessageHandlerFactory ? HttpClientExtensions.CreateHttpClientHandler(httpClientName) : null;
+        }
+
+        private void Authorize(HttpRequestMessage request)
+        {
+            switch (Credentials)
+            {
+                case NetworkCredential networkCredential:
+                    {
+                        string authString = !string.IsNullOrEmpty(networkCredential.Domain)
+                            ? networkCredential.Domain + "\\" + networkCredential.UserName + ":" + networkCredential.Password
+                            : networkCredential.UserName + ":" + networkCredential.Password;
+
+                        string base64AuthString = Convert.ToBase64String(Encoding.UTF8.GetBytes(authString));
+
+                        request.Headers.Authorization = new AuthenticationHeaderValue("Basic", base64AuthString);
+                    }
+                    break;
+
+                case null:
+                    break;
+
+                default:
+                    throw new AuthenticationException($"Credentials of type '{Credentials.GetType().Name}' is not supported");
+            }
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var cookieHeader = CookieContainer?.GetCookieHeader(request.RequestUri);
+            if (!string.IsNullOrEmpty(cookieHeader))
+            {
+                request.Headers.Add(HeaderNames.Cookie, cookieHeader);
+            }
+
+            Authorize(request);
+
+            HttpResponseMessage response = null;
+            if (null != InnerHandler)
+            {
+                response = await base.SendAsync(request, cancellationToken);
+            }
+            else if (null != httpMessageHandler)
+            {
+                var delegatingHandler = new SimpleHttpMessageHandlerProxy(httpMessageHandler);
+                response = await delegatingHandler.SendImplAsync(request, cancellationToken);
+
+                // SB: special case for Negotiate authentication, we can't reuse named HttpMessageHandler,
+                // since this type of auth requires credentials to be set on handler level, after that
+                // handler can't be reused with another credentials
+                if (response.StatusCode == HttpStatusCode.Unauthorized && Credentials is NetworkCredential && response.Headers?.WwwAuthenticate?.Count > 0)

```

Comment:
It looks like .net by default allocates custom (non-polled) handler if credentials are not set, see https://github.com/dotnet/runtime/blob/ff2de363f393f2270d5752eeb3db345fcf770c5c/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L504

So may be no need to create 2 instances. But needs tests to confirm

Reply:
looks like it is needed since SocketsHttpHandler._handler initializes once. SetupHandlerChain() called only if _handler is null:
https://github.com/dotnet/runtime/blob/ff2de363f393f2270d5752eeb3db345fcf770c5c/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L578C12-L578C79